# Variant Calling
ğŸ§¬ğŸ§ğŸ§¬ğŸ§ğŸ§¬ğŸ§ğŸ§¬ğŸ§

This step does the variant calling 



## Output:
```
[OUT_DIR]
â””â”€â”€ jATG
    â””â”€â”€ [SPECIES_NAME]
        â””â”€â”€ [ASSEMBLY_ID]
            â”œâ”€â”€ 1.stats
            â”‚   â””â”€â”€ ..
            â”œâ”€â”€ 2.masking
            â”‚   â””â”€â”€ ..
            â”œâ”€â”€ 3.gc_telo
            â”‚   â””â”€â”€ ..
            â””â”€â”€ [SAMPLE_NAME]
                â””â”€â”€ 4.snp_calling
                    â”œâ”€â”€ 0_idx
                    â”‚   â””â”€â”€ ..
                    â”œâ”€â”€ 1_BAMs
                    â”‚   â”œâ”€â”€ ..
                    â”‚   â”œâ”€â”€ eval ### BAM-based metrics
                    â”‚   â”‚   â”œâ”€â”€ <..>.markdup_metrics.txt ### dups metrics
                    â”‚   â”‚   â”œâ”€â”€ <..>.merged_MeanCov ### mean coverage
                    â”‚   â”‚   â”œâ”€â”€ <..>.merged_RefCov.md ### markdown table for coverage
                    â”‚   â”‚   â””â”€â”€ <..>.rmd_PrimAligRead ### Number of primary aligned reads [CHECK THIS, it's counting dups?]
                    â”‚   â”œâ”€â”€ <..>.merged.bam
                    â”‚   â””â”€â”€ <..>.merged.bam.bai
                    â””â”€â”€ 2_VCFs
                        â”œâ”€â”€ filt ### VCF-based metrics and filtered VCFs
                        â”‚   â”œâ”€â”€ <..>.PASS.lqual ### [site quality]
                        â”‚   â”œâ”€â”€ <..>.PASS.ldepth.mean  ### [mean coverage depth per site]
                        â”‚   â”œâ”€â”€ <..>.PASS.vcf_meanCov ### [Mean and SD coverage]
                        â”‚   â”œâ”€â”€ <..>.PASS_snps_amount ### [Amount of total and filtered SNPs]
                        â”‚   â”œâ”€â”€ <..>.PASS_filtered.vcf.gz ### [filtered VCF]
                        â”‚   â”œâ”€â”€ <..>.PASS_filtered.vcf.gz.csi
                        â”‚   â”œâ”€â”€ <..>.PASS.masked_filtered.vcf.gz ### [filtered & masked VCF]
                        â”‚   â””â”€â”€ <..>.PASS.masked_filtered.vcf.gz.csi
                        â”œâ”€â”€ <..>.g.vcf.gz
                        â”œâ”€â”€ <..>.g.vcf.gz.tbi
                        â”œâ”€â”€ <..>.FULL.g.vcf.gz  ### [base pair resolution gVCF]
                        â”œâ”€â”€ <..>.FULL.g.vcf.gz.csi
                        â”œâ”€â”€ <..>.vcf.gz
                        â”œâ”€â”€ <..>.vcf.gz.tbi
                        â”œâ”€â”€ <..>.PASS.vcf.gz ### [unfiltered VCF]
                        â”œâ”€â”€ <..>.PASS.vcf.gz.csi
                        â””â”€â”€ <..>.visual_report.html
```

### How to run?

Requirements:
* [Slurm](https://slurm.schedmd.com)
* [Conda](https://docs.conda.io)
* [Singularity](https://sylabs.io/guides/3.0/user-guide/index.html)
* PacBio HiFi trimmed reads should end with (trim\*fq, or fq or trim\*fastq or fastq).gz
* Illumina paired-end trimmed reads should end with (1.trim\*fq, or 1.fq or 1.trim\*fastq or 1.fastq).gz

1) Edit `1.calling_variables.cnf` file with the respective paths, values and parameters.

2) Install needed software with: `bash 2.install_dependencies.sh`

3) Run the variant calling pipeline in _Slurm_ with: `bash 3.Run_DeepVariant_calling.sh`

\*) The pipeline is configured to use 24 CPUs and 64 GB of RAM
