# 1. Quick overview
ğŸ§¬âš¡ï¸ğŸ§¬âš¡ï¸ğŸ§¬âš¡ï¸ğŸ§¬âš¡ï¸ğŸ§¬âš¡ï¸ğŸ§¬

This first analysis aims to generate a rapid overview of the genome assembly and produce outputs that come in handy posteriorly. You can provide a reference or close-related species for some analysis.

## Output:
```
[OUT_DIR]
â””â”€â”€ jATG
    â””â”€â”€ [SPECIES_NAME]
        â””â”€â”€ [ASSEMBLY_ID]
            â””â”€â”€ 1.overview
                â”œâ”€â”€ <..>_shortStats.md                     # very brief stats of the assembly
                â”œâ”€â”€ full_fasta_header                      # complete header of each scaffold
                â”œâ”€â”€ main_scaffolds                         # list with scaffolds longer than 5 Mbp
                â”œâ”€â”€ number_lengths_GC_Ns                   # table with name, length (decreasing order), GC rate, % of Ns of each scaffold
                â”œâ”€â”€ dotplot.<..>
                â”‚   â”œâ”€â”€ <..>_main_scaff.paf                # alignment between the main scaffolds of the assembly and a reference (if provided)
                â”‚   â”œâ”€â”€ <..>.html                          # interactive DotPlot between the assembly and a reference (if provided)
                â”‚   â””â”€â”€ <..>.png                           # png of the DotPlot between the assembly and a reference (if provided)
                â”œâ”€â”€ GC_Telo
                â”‚   â”œâ”€â”€ ..
                â”‚   â”œâ”€â”€ <..>_GC.tab                        # GC% across the assembly
                â”‚   â”œâ”€â”€ <..>_telomeric_repeat_windows.tsv  # telomeric signal across the assembly
                â”‚   â”œâ”€â”€ <..>_gaps.bed                      # position of the gaps across the assembly
                â”‚   â””â”€â”€ plots
                â”‚       â”œâ”€â”€ COMBINED_GAPS_GC_TELO.svg      # plot showing GC%, telomeric signal and gaps in all chromosomes
                â”‚       â”œâ”€â”€ <..>_GAPS_GC_TELO.svg          # plot showing GC%, telomeric signal and gaps in each chromosome
                â”‚       â””â”€â”€ ..
                â””â”€â”€ sexChr.<..>
                    â”œâ”€â”€ <..>_sexChrSeqs.faa                # sex chromosomes-linked genes sequences identified from a reference (if provided)
                    â”œâ”€â”€ <..>_sexChrGenes.md                # table of sex chromosomes-linked genes IDs identified from a reference (if provided)
                    â”œâ”€â”€ sexChr.paf                         # alignment between sex-chromosome linked genes from reference (if provided) and the assembly
                    â””â”€â”€ sexChr_evidence                    # list with number of hits for sex-chromosome linked genes between reference (if provided) and each scaffold
```

### How to run?

1) Edit `1.stats_variables.cnf` file with the respective paths, values and parameters.

2) Check if the provided reference of close-related species contains sex-chromosome relevant data: `bash 2.reference_check.sh`

3) Run the analysis based on the inputs provided: `bash 3.Run_stats.sh`

\*) It uses up to 6 cpus and 18 Gb of RAM


### About the analysis and setting of the parameters:
