# Genome-wide GC and Telomeric Regions Identification
ğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬ğŸ“ˆğŸ§¬ğŸ”šğŸ§¬

This first analysis aims to generate a rapid overview of the genome assembly and produce outputs that come in handy posteriorly.

## Output:
```
[OUT_DIR]
â””â”€â”€ jATG
    â””â”€â”€ [SPECIES_NAME]
        â””â”€â”€ [ASSEMBLY_ID]
            â””â”€â”€ 1.stats
                â”œâ”€â”€ <..>_shortStats.tsv     # very brief stats of the assembly
                â”œâ”€â”€ full_fasta_header       # complete header of each scaffold
                â”œâ”€â”€ main_scaffolds          # list with scaffolds longer than 5 Mbp
                â”œâ”€â”€ number_lengths_GC_Ns    # table with number of scaffold, name, length (decreasing order), GC rate, % of Ns 
                â”œâ”€â”€ <..>_sequences.fasta    # sex chromosomes linked genes sequences identified for the species
                â”œâ”€â”€ <..>_genes.txt          # sex chromosomes linked genes IDs
                â”œâ”€â”€ <..>.blast              # table with matches between sex chromosome linked genes and the assembly
                â”œâ”€â”€ <..>.SYNTENYTAB.tsv     # synteny between the assembly and a reference (if provided)
                â”œâ”€â”€ <..>.png                # DotPlot between the assembly and a reference (if provided)
                â””â”€â”€ ..

```

### How to run?

Requirements:
* [Slurm](https://slurm.schedmd.com)
* [Conda](https://docs.conda.io)


1) Edit `1.stats_variables.cnf` file with the respective paths, values and parameters.

2) Install needed software with: `bash 2.install_stats_dependencies.sh`

3) Run the masking pipeline in _Slurm_ with: `bash 3.Run_stats.sh`

\*) It uses up to 8 cpus and 16 Gb of RAM




##### The GC content may be correlated with underlying rates of recombination through GC-biased gene conversion.


1. Edit 1.gc_variables.cnf file with the respective paths, values and parameters.


2. Install needed software with: bash 2.install_gc_dependencies.sh


3. Run the GC pipeline with: bash 3.runGC.sh
